tools/jit/templates/aten_dispatch.cpp

```diff
@@ -20,19 +20,75 @@ using operator_constructor = std::function<TensorOp(jit::Node*)>;
 
 namespace {
 
-void pack_list(std::vector<Tensor> & outputs, Tensor v) { outputs.push_back(v); }
-void pack_list(std::vector<Tensor> & outputs, Scalar v) { outputs.push_back(v.toTensor()); }
-void pack_list(std::vector<Tensor> & outputs, const std::vector<Tensor> & t) {
-  outputs.insert(outputs.end(), t.begin(), t.end());
+// a temporary Tensor that does not alter the refcount of impl on
+// acquisition or release, avoids any refcounting in dispatch functions
+struct TensorTemporary {
+  explicit TensorTemporary(at::Retainable * impl)
+  : temp(static_cast<at::TensorImpl*>(impl), false /* do not retain*/) {}
+  const at::Tensor & value() {
+    return temp;
+  }
+  ~TensorTemporary() {
+    // don't reduce the refcount on deletion
+    temp.detach();
+  }
+private:
+  at::Tensor temp;
+};
+
+// same list of Tensors that does not alter the refcount on acquisition or
+// release of the refcount temporaries, only used rarely (e.g. for cat)
+struct TensorTemporaryList {
+  explicit TensorTemporaryList(const list_of_retainable & ts) {
+    tensors.reserve(ts.size());
+    for(auto & t : ts) {
+      tensors.push_back(at::Tensor(static_cast<at::TensorImpl*>(t), false /*do not retain*/));
+    }
+  }
+  // TensorTemporaryList only exposes a TensorList,
+  // not its underlying std::vector<at::Tensor>.
+  // This ArrayRef has the desired semantics: if you get out an at::Tensor from it,
+  // the refcount is bumped;
+  // if you take a reference, it is only guaranteed to stay live as long as the ArrayRef is live,
+  operator TensorList() const {
+    return tensors;
+  }
+  ~TensorTemporaryList() {
+    // we didnt retain the tensors when we created the list
+    // so make sure we don't release them when we free it
+    for(auto & t : tensors) {
+      t.detach();
+    }
+  }
+private:
+  std::vector<at::Tensor> tensors;
+};
+
+using list_of_retainable = std::vector<at::Retainable*>;
+
+// pack list takes the return values of aten functions and puts them into a
+// refcounted list. Each pack_list variant takes a Tensor by value, ensuring
+// it has a owning reference and then that reference is stolen ad added to the
+// list_of_retainable output list.
+// pack_list never operates on tensor temporaries.
+void pack_list(list_of_retainable & outputs, Tensor v) { outputs.push_back(toRetainableSteal(std::move(v))); }
```

Comment:
It's still a bit ugly that `list_of_retainable` doesn't manage the lifetime of the pointers it holds - some code might look as if it should free memory when an error happens, while in reality you have these unsafe moments before you plug these pointers back to registers, when any exceptions will basically leak memory.

Can't we just have an `std::vector<at::Retainable*>` for inputs, because they are temporaries that don't bump refcounts, and `std::vector<std::unique_ptr<at::Retainable>>` (with a custom deleter) to ensure we will never leak outputs?

Reply:
I tried using std::vectors of unique_ptrs for the register list and for outputs. It doesn't really work: for outputs, you add a whole useless loop during the destruction of the list to check for destructors that will never run. for the register list, it can't be a trivial vector<unique_ptr> list because it has to support a copy constructor. It can be a new container that uses a unique_ptr underneath, but that container is almost as long as the implementation of the current container anyway with little added benefit. Working around this stuff just doesn't seem worth the effort.

Reply:
Ok, let's leave it for later :+1: